cmd/tnf/addclaim/addclaim.go

```diff
@@ -27,12 +29,268 @@ var (
+		} `json:"nodes"`
+	} `json:"claim"`
+}
+
+type HwInfo struct {
+	Claim struct {
+		Nodes struct {
+			NodesHwInfo map[string]interface{} `json:"nodesHwInfo"`
+		} `json:"nodes"`
+	} `json:"claim"`
+}
+
+type RawResult struct {
+	Claim struct {
+		RawResults struct {
+			Cnfcertificationtest struct {
+				Testsuites struct {
+					Testsuite struct {
+						Testcase []testCase `json:"testcase"`
+					} `json:"testsuite"`
+				} `json:"testsuites"`
+			} `json:"cnf-certification-test"`
+		} `json:"rawResults"`
+	} `json:"claim"`
+}
+
+func claimCompare(cmd *cobra.Command, args []string) error {
+	claimFileTextPtr := Claim1
+
+	claimFileTextPtr2 := Claim2
+	err := claimCompareFilesfunc(claimFileTextPtr, claimFileTextPtr2)
+	if err != nil {
+		log.Fatalf("Error rclaimCompareFilesfunc :%v", err)
+	}
+	return nil
+}
+
+type testCase struct {
+	Name   string `json:"-name"`
+	Status string `json:"-status"`
+}
+
+func claimCompareFilesfunc(claim1, claim2 string) error {
+	// readfiles
+	calimdata1, err := os.ReadFile(claim1)
+	if err != nil {
+		log.Fatalf("Error reading claim1 file:%v", err)
+	}
+	calimdata2, err2 := os.ReadFile(claim2)
+	if err != nil {
+		log.Fatalf("Error reading claim2 file :%v", err2)
+	}
+	// unmarshal the files
+	cni1, hwinfo1, rawResult1, err := unmarshalClaimFile(calimdata1)
+	if err != nil {
+		log.Fatalf("Error in unmarshal claim1 file  :%v", err)
+		return err
+	}
+	cni2, hwinfo2, rawResult2, err := unmarshalClaimFile(calimdata2)
+	if err != nil {
+		log.Fatalf("Error in unmarshal cliam2 file  :%v", err)
+		return err
+	}
+	// compares function
+	compare2cni(cni1.Claim.Nodes.CniPlugins, cni2.Claim.Nodes.CniPlugins)
+	compare2Hwinfo(hwinfo1.Claim.Nodes.NodesHwInfo, hwinfo2.Claim.Nodes.NodesHwInfo)
+
+	slist, r, r2 := compare2TestCaseResults(rawResult1.Claim.RawResults.Cnfcertificationtest.Testsuites.Testsuite.Testcase,
+		rawResult2.Claim.RawResults.Cnfcertificationtest.Testsuites.Testsuite.Testcase)
+	log.Info("claim1 and claim2 has diff RawResults ", slist)
+	log.Info("test name that claim1 has but claim 2 dont has", r)
+	log.Info("test name that claim2 has but claim 1 dont has", r2)
+
+	return nil
+}
+
+func unmarshalClaimFile(calimdata []byte) (Cni, HwInfo, RawResult, error) {
+	var cni Cni
+	var hwinfo HwInfo
+	var rawResult RawResult
+
+	errcni := json.Unmarshal(calimdata, &cni)
+	if errcni != nil {
+		log.Fatalf("Error in unmarshal the cni from claim2 file  :%v", errcni)
+		return cni, hwinfo, rawResult, errcni
+	}
+	// csi
+	var csi Csi
+	errcsi := json.Unmarshal(calimdata, &csi)
+	if errcsi != nil {
+		log.Fatalf("Error in unmarshal the csi from claim1 file  :%v", errcsi)
+		return cni, hwinfo, rawResult, errcsi
+	}
+
+	// HwInfo
+	errhwinfo := json.Unmarshal(calimdata, &hwinfo)
+	if errhwinfo != nil {
+		log.Fatalf("Error in unmarshal the hwinfo from claim1 file  :%v", errhwinfo)
+		return cni, hwinfo, rawResult, errhwinfo
+	}
+
+	// rawResult
+	errrawResult := json.Unmarshal(calimdata, &rawResult)
```

Comment:
Why are you unmarshaling the same file three times? If you create the whole struct for the claim file you only need to do one unmarshaling. This is an example of the whole claim file struct that you need for the checks you're doing at the moment:
```
type ClaimFile struct {
	Claim struct {
		Nodes struct {
			CniPlugins map[string]cnistruct `json:"cniPlugins"`
			NodesHwInfo map[string]interface{} `json:"nodesHwInfo"`
			CsiDriver interface{} `json:"csiDriver"`
		} `json:"nodes"`

		RawResults struct {
			Cnfcertificationtest struct {
				Testsuites struct {
					Testsuite struct {
						Testcase testcase `json:"testcase"`
					} `json:"testsuite"`
				} `json:"testsuites"`
			} `json:"cnf-certification-test"`
		} `json:"rawResults"`
	} `json:"claim"`
}
```

Reply:
agree i fixed